### ISSUE TITLE
Annotation required for `eq.substr`

### ISSUE THREAD

**Issue by ranweiler** [leanprover/lean3#1718]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean/issues).
        * Specifically, check out the [wishlist](https://github.com/leanprover/lean/issues?q=is%3Aissue+is%3Aopen+label%3AI-wishlist), open [RFCs](https://github.com/leanprover/lean/issues?q=is%3Aissue+is%3Aopen+label%3ARFC),
        or [feature requests](https://github.com/leanprover/lean/issues?q=is%3Aissue+is%3Aopen+label%3AFeature).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

Type checking fails when attempting to apply `eq.substr `, even though inlining its body succeeds.

### Steps to Reproduce

1. Consider the following development in a new file, say, `test.lean`:
```lean
theorem my_theorem : ∀ (a b c d : ℕ),
  d = c → a = b * c → a = b * d :=
take a b c d,
assume h₀ : d = c,
assume h₁ : a = b * c,
eq.subst (eq.symm h₀) h₁
```
This type checks, and proof succeeds.

2. Replace the final line of the proof with `eq.substr h₀ h₁`.
3. Attempt to check the proof.

**Expected behavior:**
I expect to the use of `eq.substr` to type check.

**Actual behavior:**
The new final line fails to type check with the error:
```
/home/joe/src/nt-proofs/test.lean:20:0: error: type mismatch at application
  eq.substr h₀
term
  h₀
has type
  d = c
but is expected to have type
  b * d = ?m_1
```
If we add `set_option pp.all true` to the top of the file, the error message is:
```
/home/joe/src/nt-proofs/test.lean:22:0: error: type mismatch at application
  @eq.substr.{1} nat (@eq.{1} nat a) ?m_1 (@has_mul.mul.{0} nat nat.has_mul b d) h₀
term
  h₀
has type
  @eq.{1} nat d c
but is expected to have type
  @eq.{1} nat (@has_mul.mul.{0} nat nat.has_mul b d) ?m_1
```

**Reproduces how often:**
100% reproducible.

### Versions

Platform: Debian Linux 9 "Stretch", `Linux 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u1 (2017-06-18) x86_64 GNU/Linux`
Lean version: `Lean (version 3.2.1, commit f7fe2a775c14, Release)`

### Additional Information

I took a look at the definition of `eq.substr` in `init/logic.lean` and compared it to the definition of `eq.subst` in `init/core.lean`. The definition of `eq.subst` has the annotation `@[elab_as_eliminator, subst]`. If I make a copy of `eq.substr` in my test file, and add the (smaller) annotation `@[elab_as_eliminator]`, then the final line type checks.

I poked around `frontends/elaborator.cpp` to see what this does (since it isn't seem to be documented [quite yet](https://github.com/leanprover/reference/tree/master/Attributes)), but I don't yet understand enough of Lean's internals to be able to confidently suggest any of: (1) this is expected behavior, (2) this is a bug, or (3) the `eq.substr` definition simply should have the annotation added.

I'm still working through the "Theorem Proving in Lean" tutorial, but I've searched the docs extensively, and am having trouble meaningfully diagnosing what is going on myself. I'd love to have a sense of the following:
1. Is there a way to manually/locally accomplish what is accomplished by the `elab_as_eliminator` annotation on the `eq.subst` definition? In other words, is there a way I could invoke `eq.substr` in the above example that would succeed, via perhaps a use of `@eq.substr` or `eq.rec_on`?
1. Adding a file-local attribute like `local attribute [elab_as_eliminator] eq.substr` doesn't seem to have any effect (nor an error message or warning). Is this expected?
1. Even having found a "fix", I don't yet understand how to relate the error message to what was going wrong. I instead relied on trial-and-error and browsing source code. What surprised me is that the error seemed to be about the type of `h₀`, the simple equality `d = c`, and not `h₁`, the argument for which Lean needs to infer the implicit predicate argument. In a case like this, should I, as a user, be able to work from the existing error message to the understanding that e.g. some sort of extra elimination is needed? If anything, it seemed like the error message was expecting the _wrong_ type. I'm guessing this has to do with elaborator internals, but maybe there's something simpler I'm currently ignorant of. Any pointers on how to think about this specific problem (and to better use the existing error messages) would be much appreciated!

**Comment by leodemoura** [comment]
The following message from `lean-user` explains the meaning of the attribute `[elab_as_eliminator]`.

> Is there a way to manually/locally accomplish what is accomplished by the elab_as_eliminator annotation on the eq.subst definition?

Most users use tactics for applying substitutions.
Here are a few examples
```lean
example : ∀ (a b c d : ℕ), d = c → a = b * c → a = b * d :=
take a b c d, assume h₀ : d = c, assume h₁ : a = b * c,
by subst h₀; exact h₁

example : ∀ (a b c d : ℕ), d = c → a = b * c → a = b * d :=
take a b c d, assume h₀ : d = c, assume h₁ : a = b * c,
by simp [h₀, h₁]

example : ∀ (a b c d : ℕ), d = c → a = b * c → a = b * d :=
take a b c d, assume h₀ : d = c, assume h₁ : a = b * c,
by rw [h₀]; assumption

example : ∀ (a b c d : ℕ), d = c → a = b * c → a = b * d :=
take a b c d, assume h₀ : d = c, assume h₁ : a = b * c,
by simp_all

example : ∀ (a b c d : ℕ), d = c → a = b * c → a = b * d :=
take a b c d, assume h₀ : d = c, assume h₁ : a = b * c,
by simph

example : ∀ (a b c d : ℕ), d = c → a = b * c → a = b * d :=
take a b c d, assume h₀ : d = c, assume h₁ : a = b * c,
by cc
```

> Adding a file-local attribute like local attribute [elab_as_eliminator] eq.substr doesn't seem to have any effect (nor an error message or warning). Is this expected?

I can't reproduce. The following works for me (I'm using the `master` branch).

```lean
local attribute [elab_as_eliminator] eq.substr

theorem my_theorem : ∀ (a b c d : ℕ),
  d = c → a = b * c → a = b * d :=
take a b c d,
assume h₀ : d = c,
assume h₁ : a = b * c,
eq.substr h₀ h₁
```

> Even having found a "fix", I don't yet understand how to relate the error message to what was going wrong. I instead relied on trial-and-error and browsing source code. What surprised me is that the error seemed to be about the type of h₀, the simple equality d = c, and not h₁, the argument for which Lean needs to infer the implicit predicate argument. In a case like this, should I, as a user, be able to work from the existing error message to the understanding that e.g. some sort of extra elimination is needed?

For definitions such as `eq.substr`, we need higher-order unification. The default strategy for elaborating applications doesn't use it. The unification problem `?p ?a =?= (a = b * c)`, where `?p` and `?a` represent `eq.substr` implicit arguments, is solved using first-order unification, and the solution is not useful: `?p := @eq nat a` and `?a := b * c`, and the weird error message is produced from there.

**Comment by leodemoura** [comment]
@avigad Chapter 4 at (https://leanprover.github.io/theorem_proving_in_lean/#04_Quantifiers_and_Equality.html) still uses `eq.subst`. Should we remove it? The same chapter is also describing how to use `rw` and `simp`.
Another option is to explain it requires higher-order unification, describe why it is expensive and unreliable, etc. It might be a good place for describing the elaboration strategies too.

**Comment by ranweiler** [comment]
@leodemoura, thanks so much for the comprehensive response!

> The following message from lean-user explains the meaning of the attribute [elab_as_eliminator].

I was unaware of `lean-user`— thank you for that pointer, I'll be sure to consult that list's archives for future questions.

> I can't reproduce.

I am also unable to this reproduce it this morning. I'm going to assume that was a typo on my end.

**Comment by leodemoura** [comment]
@gebner @kha @jroesch We can try to improve the error message by checking whether the function has implicit arguments that require higher-order unification (e.g.,  `p : α → Prop` in the example above), and if does we append extra information in the error message (e.g., `function 'eq.substr' require higher-order unfication ...`). We may also try the `elab_as_eliminator` strategy, and if it succeeds, we also include the message `consider marking 'eq.substr' with the attribute '[elab_as_eliminator]'`). What do you think?

**Comment by ranweiler** [comment]
FWIW, even though I assume that in most scenarios I want to use tactics for much of the work in my proof scripts, as a new user, do I find it very helpful to have the documentation include more "laborious"/manual examples using e.g. `eq.subst`. Official documentation on the elaborator strategies would be very helpful.

**Comment by avigad** [comment]
I will explain the elaboration strategies. Maybe I will do it in a later, more advanced chapter, but add a forward reference in Chapter 4 and warn the user of potential problems.

By the way, yesterday I expanded the discussion of `rewrite` and `simp` in Chapter 5. I will also add a mention of `cc`.

**Comment by Kha** [comment]
@leodemoura I don't expect this to be a common problem, but a better error message still sounds like a good idea.

**Comment by ranweiler** [comment]
Oh, I will add one suggestion. Since what I reported is expected behavior, consider adding the `elab_as_eliminator` attribute— and possibly the `subst` attribute— to the definition of `eq.substr`, so its annotations match that of `eq.subst`.

The important thing I learned from this issue is that the elaboration strategy is in some sense part of the "API" of a theorem. My confusion came from ignorance of this. Adding `eq.subst`'s annotations to `eq.substr` would present a more consistent API, which seems appropriate if `eq.substr` is truly just meant to be a convenience lemma.
